common/sampling.cpp

```diff
@@ -400,7 +400,7 @@ static llama_token_data_array llama_sampling_prepare_impl(
     llama_token_data_array cur_p = { cur.data(), cur.size(), false };
 
     // apply penalties
-    const auto& penalty_tokens = params.use_penalty_prompt_tokens ? params.penalty_prompt_tokens : prev;
+    const auto& penalty_tokens = params.use_penalty_prompt_tokens ? params.penalty_prompt_tokens : prev.to_vector();
```

Comment:
Should think of a way to avoid the `to_vector()` due to performance considerations

Reply:
> Should think of a way to avoid the `to_vector()` due to performance considerations

I think one way to avoid vector copy is here we can pass `penalty_prompt_tokens` vector and `start index` to `llama_sample_repetition_penalties`, and then traverse `penalty_last_n` elements from the vector in it, which will avoid the copy.  For example 

```cpp
void llama_sample_repetition_penalties(
            struct llama_context * ctx,
          llama_token_data_array * candidates,
                         // const llama_token * last_tokens,
                          const vector<llama_token>& penalty_tokens,
                          size_t start_index, // .size() - penalty_tokens_used_size or (prev.first +.size() - penalty_tokens_used_size) % prev.capacity if ring buffer
                          size_t   penalty_last_n,
                           float   penalty_repeat,
                           float   penalty_freq,
                           float   penalty_present);
```

what do you think?

Reply:
After the sampling refactoring, the `common/sampling.h/.cpp` stuff will be moved to `llama-sampling.cpp` and the API call will become simply:

```c
void llama_sampling_repetition_penalties(
           struct llama_sampling * ctx,
          llama_token_data_array * candidates);
```

All the penalty related information (together with ring buffer with the previous tokens) will be inside the `llama_sampling` object and we can handle it there. So for now, we can just resolve the conflict and merge and later I'll avoid the `to_vector()`